Add template to RPC commands documentation

lucad70 opened the pull request:
### What is the purpose of this pull request?

- [ ] Bug fix
- [X] Documentation update
- [ ] New feature
- [ ] Test
- [ ] Other: <!-- Please describe it -->

### Which crates are being modified?

- [ ] floresta-chain
- [x] floresta-cli
- [ ] floresta-common
- [ ] floresta-compact-filters
- [ ] floresta-electrum
- [ ] floresta-watch-only
- [ ] floresta-wire
- [ ] floresta
- [ ] florestad
- [X] Other: CONTRIBUTING.md <!-- Please describe it -->

### Description
In other to better tackle RPCs, I suggest following this template on the CONTRIBUTING.md as discussed on #453 .

<!-- Describe the purpose of this PR, what's being added and/or fixed. If there's an open issue for it, link it here -->

### Notes to the reviewers
When working on this I tried different approachs:
1. Creating a custom `man` command to generate a pretty documentation
2. Using colored and implementing a style for CLI help and long description
3. Just defining the usage on the docs itself `enum Methods` and on the trait.

1. 
![image](https://github.com/user-attachments/assets/78763892-da50-494b-aa74-82e49aa308bc)

This looks really nice, but in the longterm I believe it is going to be hindrance to keep and use. So I opted for the direct `doc` approach (No. 3). Which allows for usage of `floresta-cli help <command>` and generating corresponding docs if applied the same docs on the trait.

<img width="584" alt="Screenshot 2025-06-13 at 15 11 52" src="https://github.com/user-attachments/assets/f3b7ba66-bb88-4b6a-8457-b5253180a183" />
<img width="781" alt="Screenshot 2025-06-13 at 15 16 07" src="https://github.com/user-attachments/assets/878a5c58-4068-44e5-8711-21101a005358" />

<!-- In this section you can include notes directed to the reviewers, like explaining why some parts of the PR were done in a specific way -->

### Author Checklist

<!-- Feel free to remove this section once you've confirmed all items -->

- [X] I've followed the [contribution guidelines](https://github.com/vinteumorg/Floresta/blob/master/CONTRIBUTING.md)
- [ ] I've verified one of the following:
  - Ran `just pcc` (recommended but slower)
  - Ran `just lint-features '-- -D warnings' && cargo test --release`
  - Confirmed CI passed on my fork
- [X] I've linked any related issue(s) in the sections above

jaoleal commented on CONTRIBUTING.md:
````diff
@@ -184,7 +184,61 @@ pub fn validate_block_time(
+///
+/// # Returns
+///
+/// JSON object containing:
+/// - **blocks**: Current block height (integer)
+///
+/// - **bestblockhash**: Hash of the current best block (string)
+///
+/// - **difficulty**: Current network difficulty (float)
+///
+/// - **verificationprogress**: Chain verification progress from 0.0 to 1.0 (float)
+///
+/// - **chainwork**: Total amount of work in the active chain (string)
+///
+/// - **warnings**: Any network warnings (string)
+///
+/// # Examples
+///
+/// ```bash
+/// floresta-cli getblockchaininfo
+///
+/// floresta-cli -n testnet getblockchaininfo
+/// ```
+///
````
I would swap these in position and use a literal json object thats the output of the example command.

Davidson-Souza commented:
Here's an idea:

We can make some markdown files, place it under `doc/rpc`, and use `#[doc = include_str!("path/to/file.md")]` inside `floresta-cli` to include them. For the terminal, we can use `mn(1)` or `info(1)` to show a pretty page. This avoids duplication, and uses well-established tools

lucad70 commented:
> We can make some markdown files, place it under `doc/rpc`, and use `#[doc = include_str!("path/to/file.md")]` inside `floresta-cli` to include them. For the terminal, we can use `mn(1)` or `info(1)` to show a pretty page. This avoids duplication, and uses well-established tools

Your suggestion is half implemented right now. I am working on a `just` recipe to generate the `man` pages.

lucad70 commented:
<img width="1223" alt="Screenshot 2025-06-16 at 13 19 19" src="https://github.com/user-attachments/assets/98cb8584-253d-4fb0-8653-3e5358e62287" />


Image of the man page of addnode generated using the `just convert-all` 

I need to fix the documention for it though. It is using the underscore

@Davidson-Souza @jaoleal

lucad70 commented:
A dev question here:

Is it better to:
- push the generated man pages inside the folder
- make the user generate them each time he wants to use it
- install it

I never used man pages, so I dont know which one is better

I also saw that there is a usual "template" people follow

https://eddieantonio.ca/blog/2015/12/18/authoring-manpages-in-markdown-with-pandoc/

https://www.linuxhowtos.org/System/creatingman.htm

However, I dont see it being useful in our case. WDYT?

Davidson-Souza commented:
> A dev question here:
> 
> Is it better to:
> 
> * push the generated man pages inside the folder
> * make the user generate them each time he wants to use it
> * install it
> 
> I never used man pages, so I dont know which one is better
> 
> I also saw that there is a usual "template" people follow
> 
> https://eddieantonio.ca/blog/2015/12/18/authoring-manpages-in-markdown-with-pandoc/
> 
> https://www.linuxhowtos.org/System/creatingman.htm
> 
> However, I dont see it being useful in our case. WDYT?

Usually they get installed along with the binaries. We don't need to keep them. If you want to make some script, leave it under `contrib/dist`. Something like `gen_manpages`.

So packagers, or our install script (#362) can use it to generate and install them.

lucad70 commented on justfile:
```diff
@@ -105,3 +105,61 @@ pcc:
+
+# Convert a single markdown file on doc/rpc/ to man page on doc/rpc_man/
+# Must have pandoc installed
+convert-single FILE:
```
Thanks. Please re-review. I just added a commit where I left the `just` recipe to make it easier and also added one script under `/contrib/dist/`. And I also added `/doc/rpc_man` to `.gitignore`

jaoleal commented on justfile:
```diff
@@ -105,3 +105,63 @@ pcc:
+# Convert all markdown files on /doc/rpc/ to man pages on /doc/rpc_man/
+# Must have pandoc installed
+# Needs sudo to overwrite existing man pages
+gen-manpages:
```
You can delegate these two now to the `gen_manpages.sh` script and reduce a command since the script now handles generating a single one and multiple.

jaoleal commented on .gitignore:
```diff
@@ -3,7 +3,7 @@
 /result
 /bin
-
+/doc/rpc_man
```
So, we will ignore the compressed `man` pages and make one generate them every time ?
if so, you need to remove the commited man pages at `rpc_man/`

jaoleal replied:
why to include both the template and the addnode one so ?

jaoleal commented on CONTRIBUTING.md:
````diff
@@ -184,7 +184,31 @@ pub fn validate_block_time(
+Please always create a new rpc documentation under the specified [directory](/doc/rpc) and implement using the following syntax on the method command definition:
+
+```rust
+#[doc = include_str!("../../../doc/rpc/command.md")]
+#[command(name = "addnode", long_about = Some(include_str!("../../../doc/rpc/command.md")),disable_help_subcommand = true)]
````
I wonder if it would be better to make these strings a constant of the commands module. WYT ?

Davidson-Souza replied:
Then we couldn't use them for other stuff.

jaoleal replied:
why ? 
```Rust
const: str ADDNODE_MD_PATH = "../../../doc/rpc/command.md";

#[doc = include_str!(ADDNODE_MD_PATH)]
```

jaoleal commented on CONTRIBUTING.md:
````diff
@@ -184,7 +184,31 @@ pub fn validate_block_time(
+fn method_name
+```
+
+We also have `man pages` that can be generated using the script [gen_manpages.sh](/contrib/dist/gen_manpages.sh) for releases/distributions or if you want to generate them locally you can also use the `just convert-all` command.
````
I would like to see a mention of these instructions in some `doc/rpc/README.md` and not only here on `CONTRIBUTING.md` its easy to miss since on superficial appearances Contributing has nothing directly to do with our rpc ways.

lucad70 replied:
I have put it in the Floresta-Cli read.me as I think is more natural for the user to find it there looking for documentation on Floresta-CLI.

jaoleal commented on CONTRIBUTING.md:
````diff
@@ -184,7 +184,31 @@ pub fn validate_block_time(
+```bash
+just gen-manpages
+# or
+chmod +x contrib/dist/gen_manpages.sh
+./contrib/dist/gen_manpages.sh
````
```suggestion
sh ./contrib/dist/gen_manpages.sh <specific_md_path> # Specifying a md file will make the script only build this single one.
```

Davidson-Souza commented on CONTRIBUTING.md:
````diff
@@ -184,7 +184,31 @@ pub fn validate_block_time(
+
+```rust
+#[doc = include_str!("../../../doc/rpc/command.md")]
+#[command(name = "addnode", long_about = Some(include_str!("../../../doc/rpc/command.md")),disable_help_subcommand = true)]
````
I think we should use the big version only for `long_about`, and add a brief introduction for the short description (I believe this is the `#[doc]` part).

Otherwise the `help` command will be huge, and will be very annoying for the case where you just want to figure out what the exact RPC name is or other quick lookups.

jaoleal replied:
Actually, we have a section that does a "short about", the Description Section of the MD files could be used right ?

Davidson-Souza replied:
No need for two md files, in the `help` (no command, just help) it should be a one-liner of what this RPC is doing. Like

```rust
/// Adds a new peer to our peer connection list
```
If you want something more detailed, you call e.g. `help addnode` and then you'll see the long about

lucad70 replied:
> I think we should use the big version only for `long_about`, and add a brief introduction for the short description (I believe this is the `#[doc]` part).

I can achieve this config with the following:

```rust
    #[doc = include_str!("../../../doc/rpc/addnode.md")]
    #[command(name = "addnode",about = "Attempts to add or remove a node from the list of addnodes", long_about = Some(include_str!("../../../doc/rpc/addnode.md")),disable_help_subcommand = true)]
    AddNode {
        node: String,
        command: AddNodeCommand,
        v2transport: Option<bool>,
    },
```

Davidson-Souza replied:
This works. But I wonder if @jaoleal's suggestion of using constants would make sense here. It would make the attribute way simpler

lucad70 replied:
Yeah, I believe it is a god suggestion. I'll try it out.

lucad70 replied:
Just a clarification, I tried using const but the arg needs to be a string literal. So it did not work as intended and I moved to the current usage with `about` and `long about`

Davidson-Souza commented on CONTRIBUTING.md:
````diff
@@ -184,7 +184,31 @@ pub fn validate_block_time(
+fn method_name
+```
+
+We also have `man pages` that can be generated using the script [gen_manpages.sh](/contrib/dist/gen_manpages.sh) for releases/distributions or if you want to generate them locally you can also use the `just convert-all` command.
````
You forgot to update the command name here.

```suggestion
We also have `man pages` that can be generated using the script [gen_manpages.sh](/contrib/dist/gen_manpages.sh) for releases/distributions or if you want to generate them locally you can also use the `just gen-manpages` command.
```

Davidson-Souza commented on contrib/dist/gen_manpages.sh:
```diff
@@ -0,0 +1,96 @@
+        -o "$TARGET_DIR/${basename}.${section}"
+
+    # Compress the man page
+    gzip -f "$TARGET_DIR/${basename}.${section}"
```
Yeah, I know it can do that. Just haven't seen this being used before. But it's fine, prob better to compress anyway.

jaoleal commented on doc/rpc/template.md:
```diff
@@ -0,0 +1,29 @@
+
+## Arguments
+
+* `param1` - Description of the first parameter (type)
```
Yep, it doesnt look like ? 😆 

Tried to follow the docs convention that core does on json objs (here)[https://developer.bitcoin.org/reference/rpc/importdescriptors.html]

maybe it would be good to wrap it with Json Typing.

```json
{
 "json": "type",
}
```

lucad70 replied:
You meant something like this?
```rust 
/// JSON object containing:
/// ```json
/// {
///   "best_block": "string",
///   "height": "u32",
///   "ibd": "bool",
///   "validated": "u32",
///   "latest_work": "string",
///   "latest_block_time": "u32",
///   "leaf_count": "u32",
///   "root_count": "u32",
///   "root_hashes": "Vec<String>",
///   "chain": "string",
///   "progress": "Option<f32>",
///   "difficulty": "u64"
/// }
/// ```
```

lucad70 replied:
I tried here and it renders like this on mn(1) if I put an empty line between fields.

![image](https://github.com/user-attachments/assets/9188cbd4-a1a0-4d22-a8ac-ce416da17506)


WDYT?

lucad70 replied:
@jaoleal @Davidson-Souza

lucad70 commented on crates/floresta-cli/README.md:
````diff
@@ -3,318 +3,23 @@
 ```
 
-Available commands:
+We also have `man pages` that can be generated using the script [gen_manpages.sh](/contrib/dist/gen_manpages.sh) for releases/distributions or if you want to generate them locally you can also use the `just convert-all` command. This will generate man pages from files at `doc/rpc/*.md` to `doc/man/*.1`. It uses the `pandoc` dependency, so please install it before running the script.
````
thanks

lucad70 commented:
To reviewers: I removed a lot of documentation on `floresta-cli` README.md as I believe it was redundant and introduced documentation overhead, making it harder to keep documentation on cli updated on everyplace. Now everything related to RPC docs should be contained in the structures and on the `doc/rpc/command_name.md` which can be converted to a manpage.

jaoleal commented on justfile:
```diff
@@ -105,3 +105,9 @@ pcc:
+# Must have pandoc installed
+# Needs sudo to overwrite existing man pages
+# Convert all markdown files on /doc/rpc/ to man pages on /doc/rpc_man/
+gen-manpages:
+    ./contrib/dist/gen_manpages.sh
```
nope.

lucad70 replied:
I had to use another syntax
```rust
gen-manpages path="":
    ./contrib/dist/gen_manpages.sh {{path}}
```

jaoleal commented:
Do you plan to already implement the removed docs in this PR ?

jaoleal commented:
IMO this regression is fine. We might update the RPCSAGA tracker with these missing MD docs ?

jaoleal commented on doc/rpc/addnode.md:
```diff
@@ -0,0 +1,35 @@
+### Ok Response
+- json null
+
+### Error Enum [`CommandError`]
+Any of the error types on `rpc_types::Error`
+
```
```suggestion
### Error Enum

Any of the error types on `rpc_types::Error`

```

jaoleal commented on doc/rpc/addnode.md:
````diff
@@ -0,0 +1,35 @@
+
+## Usage Examples
+
+* General usage:
+```bash
+addnode <ip:[port]> <add|remove|onetry> [true|false]
+```
````
```suggestion
* General usage:
```bash
floresta-cli addnode <ip:[port]> <add|remove|onetry> [true|false]
```
```

jaoleal commented on doc/rpc/addnode.md:
````diff
@@ -0,0 +1,35 @@
+floresta-cli addnode 192.168.0.1 onetry false
+```
+
+## Notes
+- Will use v2transport if available unless the param is specified to `false`
````
empty line after Notes

jaoleal commented on doc/rpc/template.md:
```diff
@@ -0,0 +1,46 @@
+
+## Returns
+
+### Ok Response
+- `field1`- Description of return field (type)
+- `field2` - Description of another return field (type)
```
```suggestion
### Ok Response

- `field1`- Description of return field (type)
- `field2` - Description of another return field (type)
```

lucad70 commented on doc/rpc/addnode.md:
```diff
@@ -0,0 +1,35 @@
+Attempts to add or remove a node from the addnode list.
+
+## Arguments
+* `node` - The address of the peer to connect to;
```
I just pushed a modification to it. Also to the template, putting the `(type, required)` before the description

lucad70 replied:
Can I mark this as resolved? Or is there anything I should change?

lucad70 commented:
> You've removed the stub for `addnode.1.gz`?

Yes, I left only a stub for template.1 saying to check the instructions

Davidson-Souza commented:
> Yes, I left only a stub for template.1 saying to check the instructions

But the goal with them is to show the final structure after generation. So it doesn't make much sense to have only one of them. Actually, the template isn't generated by the script AFAICU, so it shouldn't be there?

lucad70 commented:
> > Yes, I left only a stub for template.1 saying to check the instructions
> 
> 
> 
> But the goal with them is to show the final structure after generation. So it doesn't make much sense to have only one of them. Actually, the template isn't generated by the script AFAICU, so it shouldn't be there?

Hmmm, ok. That makes sense as well. I thought about leaving the template.1 there if someone tried to type `man template.1` he/she would see instructions.

I will update them to leave only the template.1 as a stub, as the user can generate the others locally quite easily. WDYT?

Davidson-Souza commented:
> Hmmm, ok. That makes sense as well. I thought about leaving the template.1 there if someone tried to type man template.1 he/she would see instructions.

That information is already in the `README.md`. Not sure if any user would call `man` before reading the readme. This is for automated tools to figure out which files they need, and to debug mistakes (you'll see the stubs rather than the actual man if you fail to create them in your installation). Either add the `addnode.1` (and not the template, as it doesn't get generated), or remove them altogether. Maybe removing is fine, since you can figure that out by looking at the md file names?

lucad70 commented:
> That information is already in the `README.md`. Not sure if any user would call `man` before reading the readme. This is for automated tools to figure out which files they need, and to debug mistakes (you'll see the stubs rather than the actual man if you fail to create them in your installation). Either add the `addnode.1` (and not the template, as it doesn't get generated), or remove them altogether. Maybe removing is fine, since you can figure that out by looking at the md file names?

My rationale was to follow the `bitcoin-core man` [implementation](https://github.com/bitcoin/bitcoin/tree/master/doc/man).

In there, they leave a placeholder file with the path to the instructions.

lucad70 commented:
> My original comment was about making it exactly like core's, by having a stub for each generated file, but with instructions in them (the text can be the same).

In this sense, I believe leaving it empty, just the folder. Is this ok? I believe it is better than generating all the files as placeholders as in the end we will have over 100 rpc commands I guess

Davidson-Souza commented:
> > My original comment was about making it exactly like core's, by having a stub for each generated file, but with instructions in them (the text can be the same).
> 
> In this sense, I believe leaving it empty, just the folder. Is this ok? I believe it is better than generating all the files as placeholders as in the end we will have over 100 rpc commands I guess

ok. Just leave a readme on it explaining what to do

lucad70 commented:
> ok. Just leave a readme on it explaining what to do

I have left a readme on it and also pointed all the places where I talked about generating the manpages to it.

Davidson-Souza commented on CONTRIBUTING.md:
```diff
@@ -184,9 +184,36 @@ pub fn validate_block_time(
+
+To achieve this, we use the `rustdoc` tool, which generates documentation from Rust source code comments. We also use the `clap` library to generate CLI help and usage information directly from the code.
+
+Please, always create a new rpc documentation under the specified [directory](/doc/rpc) and implement using the following syntax on the method command definition:
```
nit: `s/rpc/RPC/`

Davidson-Souza commented on crates/floresta-cli/src/main.rs:
```diff
@@ -215,25 +214,8 @@ pub enum Methods {
-    /// Usage: addnode <ip:[port]> <add|remove|onetry> [true|false]
-    #[command(name = "addnode")]
+    #[doc = include_str!("../../../doc/rpc/addnode.md")]
+    #[command(name = "addnode",about = "Attempts to add or remove a node from the list of addnodes", long_about = Some(include_str!("../../../doc/rpc/addnode.md")),disable_help_subcommand = true)]
```
Maybe something like this is clearer?

```suggestion
#[command(
                 name = "addnode",
                 about = "Attempts to add or remove a node from the list of addnodes",
                  long_about = Some(include_str!("../../../doc/rpc/addnode.md")),
                  disable_help_subcommand = true)]
```

Davidson-Souza reviewed (commented):
Two nits and I think we are ready to go

Davidson-Souza commented:
You have a conflict with `floresta-cli/README.md`. Otherwise everything is ok

Davidson-Souza reviewed (approved):
ACK [99e412c](https://github.com/vinteumorg/Floresta/pull/521/commits/99e412c1acccc8e575a5a3ba023021f5061a5fd2)
